crates/primitives/src/fork/migration.rs

```diff
@@ -0,0 +1,23 @@
+use super::SpecId;
+
+/// Defines the interface of a migration
+pub trait ForkMigration {
+    /// Invoked before a spec is activated.
+    fn pre_spec_activation(&self, spec_id: SpecId) -> anyhow::Result<()>;
+    /// Invoked after a spec is activated.
+    fn post_spec_activation(&self, spec_id: SpecId) -> anyhow::Result<()>;
+}
+
+pub struct NoOpMigration {}
+
+impl ForkMigration for NoOpMigration {
+    fn pre_spec_activation(&self, _spec_id: SpecId) -> anyhow::Result<()> {
+        // Do nothing
+        Ok(())
+    }
+
+    fn post_spec_activation(&self, _spec_id: SpecId) -> anyhow::Result<()> {
+        // Do nothing
+        Ok(())
+    }
```

Comment:
i gave this some thought, what about only having a single on activation fork?

when spec activate height is reached, function is called before everything else.